### PR TITLE
mds: try to flush the mdlog when requesting the rdlock

### PR DESCRIPTION
**Pull request by lxbsz**
After client requests' early_reply() the mdlog won't be flushed
immediately, but it will hold the write lock. So if the client
sends another request to a different MDS daemon, which then needs
to request read lock, then it maybe stuck at most for 5 seconds.
Let's try to flush the mdlog which might release the write locks
after mdlog is successfully flushed.

Fixes: https://tracker.ceph.com/issues/51069
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by lxbsz** [comment]
jenkins test make check

File: src/mds/Locker.cc [hunk 1]
```diff
@@ -4624,6 +4624,17 @@ bool Locker::simple_sync(SimpleLock *lock, bool *need_issue)
...
+      // After a client request is early replied the mdlog won't be flushed
+      // immediately, but before safe replied the request will hold the write
+      // locks. So if the client sends another request to a different MDS
+      // daemon, which then needs to request read lock from current MDS daemon,
+      // then that daemon maybe stuck at most for 5 seconds. Which will lead
+      // the client stuck at most 5 seconds.
+      //
+      // Let's try to flush the mdlog when the write lock is held, which will
+      // release the write locks after mdlog is successfully flushed.
+      mds->mdlog->flush();
```
Comment (mchangir) [inline src/mds/Locker.cc:4637]: I'm not too aware of the Locker code base.
So, I was wondering if invalidating lock caches would do away with the write lock as well.
And, so, shouldn't the log flush take place before invalidating lock caches.

Comment (lxbsz) [reply src/mds/Locker.cc:4637]: > I'm not too aware of the Locker code base.
> So, I was wondering if invalidating lock caches would do away with the write lock as well.

IMO, the simple_sync() is trying to switch the lock state to LOCK_SYNC, which should only be readable by all replicas, so if it's wrlocked we need to invalidate the lock caches before releasing the wrlocks.

> And, so, shouldn't the log flush take place before invalidating lock caches.

From the code I didn't find this is a necessary and there have many other use cases just like this.

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]
Makes sense to me -- thanks.

**Comment by batrick** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-June-26

**Pull request merged**